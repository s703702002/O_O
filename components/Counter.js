import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Counter extends Component {
  static propTypes = {
    max: PropTypes.number,
    min: PropTypes.number,
  }
  static defaultProps = {
    max: null,
    min: 0,
  };
  state = {
    count: 0,
  }
  minus = () => {
    let { count } = this.state;
    const { min } = this.props;
    count -= 1;
    if (count < min) return;
    this.setState({ count });
  }
  add = () => {
    let { count } = this.state;
    const { max } = this.props;
    count += 1;
    if (count > max) return;
    this.setState({ count });
  }
  render() {
    return (
      <div className="counter">
        <button className="btn btn-light material-icons" onClick={this.minus}>remove</button>
        <span className="count">{this.state.count}</span>
        <button className="btn btn-light material-icons" onClick={this.add}>add</button>
      </div>
    );
  }
}

export default Counter;
