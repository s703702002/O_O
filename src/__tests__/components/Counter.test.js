// import React from 'react';
// import {
//   render,
//   fireEvent,
//   cleanup,
// } from 'react-testing-library';
// import 'jest-dom/extend-expect';
// import Counter from '../../components/Counter';

// afterEach(cleanup);

// test('loads and display Counter', () => {
//   const { getByText, container, asFragment } = render(<Counter defaultValue={5} />)
//   const count = getByText('5');
//   fireEvent.click(getByText('add'))
//   expect(count.textContent).toBe('6')
// });


import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import Counter from '../../components/Counter';

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it('Counter minus work!', () => {
  // Test first render and componentDidMount
  act(() => {
    ReactDOM.render(<Counter defaultValue={5} min={4} />, container);
  });
  const minus = container.querySelector('button');
  const span = container.querySelector('.count');
  expect(span.textContent).toBe('5');

  // Test second render and componentDidUpdate
  act(() => {
    minus.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  act(() => {
    minus.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(container).toMatchSnapshot();
});

it('Counter add work!', () => {
  // Test first render and componentDidMount
  act(() => {
    ReactDOM.render(<Counter defaultValue={10} max={10} />, container);
  });
  const add = container.querySelector('button:last-child');
  const span = container.querySelector('.count');
  expect(span.textContent).toBe('10');
  
  // Test second render and componentDidUpdate
  act(() => {
    add.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  act(() => {
    add.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(container).toMatchSnapshot();
});