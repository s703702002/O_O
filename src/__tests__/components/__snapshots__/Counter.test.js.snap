// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter add work! 1`] = `
<div>
  <div
    class="counter"
  >
    <button
      class="btn btn-light material-icons"
    >
      remove
    </button>
    <span
      class="count"
    >
      10
    </span>
    <button
      class="btn btn-light material-icons"
    >
      add
    </button>
  </div>
</div>
`;

exports[`Counter minus work! 1`] = `
<div>
  <div
    class="counter"
  >
    <button
      class="btn btn-light material-icons"
    >
      remove
    </button>
    <span
      class="count"
    >
      4
    </span>
    <button
      class="btn btn-light material-icons"
    >
      add
    </button>
  </div>
</div>
`;
